subsys/bluetooth/host/gatt.c

```diff
@@ -4473,3 +4348,154 @@ SETTINGS_STATIC_HANDLER_DEFINE(bt_hash, "bt/hash", NULL, db_hash_set,
+{
+	const struct addr_with_id *addr_with_id = user_data;
+	struct _bt_gatt_ccc *ccc;
+	struct bt_gatt_ccc_cfg *cfg;
+
+	/* Check if attribute is a CCC */
+	if (attr->write != bt_gatt_attr_write_ccc) {
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	ccc = attr->user_data;
+
+	/* Check if there is a cfg for the peer */
+	cfg = ccc_find_cfg(ccc, addr_with_id->addr, addr_with_id->id);
+	if (cfg) {
+		memset(cfg, 0, sizeof(*cfg));
+	}
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static int bt_gatt_clear_ccc(u8_t id, const bt_addr_le_t *addr)
+{
+	struct addr_with_id addr_with_id = {
+		.addr = addr,
+		.id = id,
+	};
+
+	bt_gatt_foreach_attr(0x0001, 0xffff, remove_peer_from_attr,
+			     &addr_with_id);
+
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		char key[BT_SETTINGS_KEY_MAX];
+
+		if (id) {
+			char id_str[4];
+
+			u8_to_dec(id_str, sizeof(id_str), id);
+			bt_settings_encode_key(key, sizeof(key), "ccc",
+					       (bt_addr_le_t *)addr, id_str);
+		} else {
+			bt_settings_encode_key(key, sizeof(key), "ccc",
+				       (bt_addr_le_t *)addr, NULL);
+		}
+
+		return settings_delete(key);
+	}
+
+	return 0;
+}
+
+static int bt_gatt_clear_cf(u8_t id, const bt_addr_le_t *addr)
+{
+	if (IS_ENABLED(CONFIG_BT_GATT_CACHING)) {
+		struct gatt_cf_cfg *cfg;
+
+		cfg = find_cf_cfg_by_addr(addr);
+		if (cfg) {
+			clear_cf_cfg(cfg);
+		}
+
+		if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+			char key[BT_SETTINGS_KEY_MAX];
+
+			if (id) {
+				char id_str[4];
+
+				u8_to_dec(id_str, sizeof(id_str), id);
+				bt_settings_encode_key(key, sizeof(key), "cf",
+						       (bt_addr_le_t *)addr,
+						       id_str);
+			} else {
+				bt_settings_encode_key(key, sizeof(key), "cf",
+						       (bt_addr_le_t *)addr,
+						       NULL);
+			}
+
+			return settings_delete(key);
+		}
+	}
+
+	return 0;
+
+}
+
+static int sc_clear_by_addr(u8_t id, const bt_addr_le_t *addr)
+{
+	if (IS_ENABLED(CONFIG_BT_GATT_SERVICE_CHANGED)) {
+		struct gatt_sc_cfg *cfg;
+
+		cfg = find_sc_cfg(id, (bt_addr_le_t *)addr);
+		if (cfg) {
+			sc_clear(cfg);
+		}
+	}
+	return 0;
+}
+
+static void bt_gatt_clear_subscriptions(const bt_addr_le_t *addr)
+{
+	if (IS_ENABLED(CONFIG_BT_GATT_CLIENT)) {
+		struct gatt_sub *sub = NULL;
+		struct bt_gatt_subscribe_params *params, *tmp;
+		sys_snode_t *prev = NULL;
+		int i;
+
+		for (i = 0; i < ARRAY_SIZE(subscriptions); i++) {
+			if (!bt_addr_le_cmp(addr, &subscriptions[i].peer)) {
+				sub = &subscriptions[i];
+				break;
+			}
+		}
+
+		if (!sub) {
+			return;
+		}
+
+		SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&sub->list, params, tmp,
+						  node) {
+			params->value = 0U;
+			gatt_sub_remove(NULL, sub, prev, params);
+		}
+	}
+}
+
+int bt_gatt_clear(u8_t id, const bt_addr_le_t *addr)
+{
+	int err;
+
+	err = bt_gatt_clear_ccc(id, addr);
+	if (err < 0) {
+		return err;
+	}
+
+	err = sc_clear_by_addr(id, addr);
+	if (err < 0) {
+		return err;
+	}
+
+	err = bt_gatt_clear_cf(id, addr);
+	if (err < 0) {
+		return err;
+	}
+
+	bt_gatt_clear_subscriptions(addr);
```

Comment:
Why not put `if (IS_ENABLED(CONFIG_BT_GATT_CLIENT))` here instead of doing it inside the function definition. That way you don't need to indent so much code.